### PR TITLE
Introduce SignTxAsFeePayer function in KeyStore

### PR DESCRIPTION
**Pull request by aidan-kwon**
## Proposed changes

- Introduce SignTxAsFeePayer function in KeyStore
- Add tests for signing functions of KeyStore

Later, the new functionality of the node wallet and APIs will be implemented based on `SignTxAsFeePayer` supporting fee-delegation txs.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: accounts/keystore/keystore.go [hunk 1]
```diff
@@ -284,6 +284,23 @@ func (ks *KeyStore) SignTx(a accounts.Account, tx *types.Transaction, chainID *b
 	return nil, ErrChainIdNil
 }
 
+// SignTxAsFeePayer signs the given transaction as a fee payer with the requested account.
+func (ks *KeyStore) SignTxAsFeePayer(a accounts.Account, tx *types.Transaction, chainID *big.Int) (*types.Transaction, error) {
```
Comment (jeonghonah) [inline accounts/keystore/keystore.go]:
```suggestion
func (ks *KeyStore) SignTxAsFeePayer(a accounts.Account, tx *types.Transaction, id *big.Int) (*types.Transaction, error) {
```

File: accounts/keystore/keystore_wallet.go [hunk 2]
```diff
@@ -113,6 +113,22 @@ func (w *keystoreWallet) SignTx(account accounts.Account, tx *types.Transaction,
 	return w.keystore.SignTx(account, tx, chainID)
 }
 
+// SignTxAsFeePayer implements accounts.Wallet, attempting to sign the given
+// transaction as a fee payer with the given account. If the wallet does not
+// wrap this particular account, an error is returned to avoid account leakage
+// (even though in theory we may be able to sign via our shared keystore backend).
+func (w *keystoreWallet) SignTxAsFeePayer(account accounts.Account, tx *types.Transaction, chainID *big.Int) (*types.Transaction, error) {
```
Comment (jeonghonah) [inline accounts/keystore/keystore_wallet.go]:
```suggestion
func (w *keystoreWallet) SignTxAsFeePayer(a accounts.Account, tx *types.Transaction, id *big.Int) (*types.Transaction, error) {
```

File: accounts/keystore/keystore.go (hunk 1 above)
Comment (aidan-kwon) [reply accounts/keystore/keystore.go]: I can't find where your naming rule is used. The variable name was copied from `SignTx`.
https://github.com/klaytn/klaytn/blob/dfb6591884fbdc2e86d56f9efc05cd39b004ad49/accounts/keystore/keystore.go#L271

File: accounts/keystore/keystore_wallet.go (hunk 2 above)
Comment (aidan-kwon) [reply accounts/keystore/keystore_wallet.go]: I can't find where your naming rule is used. The variable name was copied from SignTx.
https://github.com/klaytn/klaytn/blob/dfb6591884fbdc2e86d56f9efc05cd39b004ad49/accounts/keystore/keystore_wallet.go#L104

**Comment by aidan-kwon** [comment]
@KimKyungup @jimni1222 Please, take a look.

File: accounts/keystore/keystore_wallet.go (hunk 2 above)
Comment (KimKyungup) [reply accounts/keystore/keystore_wallet.go]: I also prefer `chainID` then `id`. :) `a` as a `account` is ok.

File: accounts/keystore/keystore_test.go [hunk 3]
```diff
@@ -387,5 +393,172 @@ func tmpKeyStore(t *testing.T, encrypted bool) (string, *KeyStore) {
...
+
+	// generate an account used for the test
+	password := ""
+	acc, err := ks.ImportECDSA(senderPrvKey, password)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// unlock the sender account to sign
+	if err := ks.Unlock(acc, ""); err != nil {
```
Comment (KimKyungup) [inline accounts/keystore/keystore_test.go]:
```suggestion
	if err := ks.Unlock(acc, password); err != nil {
```

File: accounts/keystore/keystore_test.go [hunk 4]
```diff
@@ -387,5 +393,172 @@ func tmpKeyStore(t *testing.T, encrypted bool) (string, *KeyStore) {
...
+
+	// generate an account used for the test
+	password := ""
+	acc, err := ks.ImportECDSA(feePayerPrvKey, password)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// unlock the fee payer account to sign
+	if err := ks.Unlock(acc, ""); err != nil {
```
Comment (KimKyungup) [inline accounts/keystore/keystore_test.go]:
```suggestion
	if err := ks.Unlock(acc, password); err != nil {
```

File: accounts/keystore/keystore_test.go [hunk 5]
```diff
@@ -387,5 +393,172 @@ func tmpKeyStore(t *testing.T, encrypted bool) (string, *KeyStore) {
...
+
+	// all test fee-delegation txs have the same fee payer address derived from the following private key
+	feePayerPrvKey, _ := crypto.HexToECDSA("f45a87856d14357609ce6b99645a1a7889deafbf00848bbdace60d8cd10466fa")
+
+	// Sample transactions for the test. Each tx is represented as a JSON marshalled byte array.
+	txs := map[types.TxType]string{
+		types.TxTypeLegacyTransaction:                           "0x7b226e6f6e6365223a22307830222c226761735072696365223a223078356432316462613030222c22676173223a223078393839363830222c22746f223a22307866396661643065393462323136666166666665666239396566303263653434663939346133646538222c2276616c7565223a22307831222c22696e707574223a223078222c227369676e617475726573223a5b7b2256223a2230783235222c2252223a223078666666666635643530386235396636343337333335396137316364613661356362613639373833326538353033393035333363303463303534353632363466222c2253223a22307834313935313931366366363739653131356136366532613130616365363530363166333836663937396437613038616264333866393265333336303437353361227d5d2c2268617368223a22307838666434393838393364383031376437316635643237313432653436316636303365366538393861613432376338613736316236623539323734346361343861227d",
```
[the remark this answers is not in the document]
Comment (KimKyungup) [reply accounts/keystore/keystore_test.go]: If you insist to use generated tx to test compatibility, why don't you explain or test code how to generate this test txs. :)

File: accounts/keystore/keystore_test.go (hunk 3 above)
Comment (aidan-kwon) [reply accounts/keystore/keystore_test.go]: Thanks!

File: accounts/keystore/keystore_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (aidan-kwon) [reply accounts/keystore/keystore_test.go]: The reason why the byte format is used. 
1. We need a static expected signature to verify the functionality of the signing function. If we generate a tx on each time, the expected signature needs to be changed later when the tx format is changed or something happened in tx (This test is not for tx format, but for signing function).
I thought static values are better to test signing function itself. 
2. easy to generate (for me) and short test code

I will add a comment providing a hint to generate those byte codes.

File: blockchain/types/transaction.go [hunk 6]
```diff
@@ -43,7 +43,7 @@ var (
 	ErrInvalidTxTypeForAnchoredData   = errors.New("invalid transaction type for anchored data")
 	errLegacyTransaction              = errors.New("should not be called by a legacy transaction")
 	errNotImplementTxInternalDataFrom = errors.New("not implement TxInternalDataFrom")
-	errNotFeePayer                    = errors.New("not implement fee payer interface")
+	errNotFeePayer                    = errors.New("not a fee delegation type transaction")
```
[the remark this answers is not in the document]
Comment (aidan-kwon) [reply blockchain/types/transaction.go]: Reflected. Thank you!

File: accounts/keystore/keystore_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (KimKyungup) [reply accounts/keystore/keystore_test.go]: Yes. please provide the code to generate them. :)

[the remark this answers is not in the document]
Comment (aidan-kwon) [reply accounts/keystore/keystore_test.go]: `"TestDefaultTxsWithDefaultAccountKey" in tests package may help you to generate byte codes of all tx types.`
@KimKyungup Isn't it enough?

**Comment by aidan-kwon** [comment]
Reviewers, I refined test cases. Please take another look.

File: accounts/keystore/keystore_test.go [hunk 7]
```diff
@@ -387,5 +393,111 @@ func tmpKeyStore(t *testing.T, encrypted bool) (string, *KeyStore) {
...
+	}
+
+	return senderPrvKey, feePayerPrvKey, tx
+}
+
+// TestKeyStore_SignTx tests the tx signing function of KeyStore.
+func TestKeyStore_SignTx(t *testing.T) {
+	chainID := big.NewInt(1)
+
+	// test transactions and the private key of the sender
```
Comment (aidan-kwon) [inline accounts/keystore/keystore_test.go]:
```suggestion
	// test transaction and the private key of the sender
```

File: accounts/keystore/keystore_test.go [hunk 8]
```diff
@@ -387,5 +393,111 @@ func tmpKeyStore(t *testing.T, encrypted bool) (string, *KeyStore) {
...
+
+	// Two signing functions should return the same signature
+	assert.Equal(t, sig2, sig1)
+}
+
+// TestKeyStore_SignTxAsFeePayer tests the fee payer's tx signing function of KeyStore.
+func TestKeyStore_SignTxAsFeePayer(t *testing.T) {
+	chainID := big.NewInt(1)
+
+	// test transactions and the private key of the sender
```
Comment (aidan-kwon) [inline accounts/keystore/keystore_test.go]:
```suggestion
	// test transaction and the private key of the sender
```

File: accounts/keystore/keystore_test.go [hunk 9]
```diff
@@ -387,5 +393,111 @@ func tmpKeyStore(t *testing.T, encrypted bool) (string, *KeyStore) {
...
+	password := ""
+	acc, err := ks.ImportECDSA(senderPrvKey, password)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := ks.Unlock(acc, password); err != nil {
+		t.Fatal(err)
+	}
+
+	// get a signature signature from a signing function in KeyStore
```
Comment (jimni1222) [inline accounts/keystore/keystore_test.go]:
```suggestion
	// get a signature from a signing function in KeyStore
```

**Comment by aidan-kwon** [comment]
I rebased this PR to avoid test fail issue which is resolved in recent.

File: accounts/keystore/keystore_test.go [hunk 10]
```diff
@@ -387,5 +393,111 @@ func tmpKeyStore(t *testing.T, encrypted bool) (string, *KeyStore) {
...
```
Comment (jimni1222) [inline accounts/keystore/keystore_test.go]: It seems that the modified content is missing while rebase.

Comment (aidan-kwon) [reply accounts/keystore/keystore_test.go]: Oh! thank you. I'm gonna update them.

File: accounts/keystore/keystore_wallet.go (hunk 2 above)
Comment (jeonghonah) [reply accounts/keystore/keystore_wallet.go]: If 'account accounts.Account' is appropriate here, we'd better change `SignTxAsFeePayer()`.

Comment (aidan-kwon) [reply accounts/keystore/keystore_wallet.go]: Okay. I will change it later. We need to change the variable name in the whole file.

**Pull request merged**